src/dbus/dbus_thread_object.cpp

```diff
@@ -0,0 +1,558 @@
+{
+    auto                 threadHelper = mNcp->GetThreadHelper();
+    std::string          name;
+    uint16_t             panid;
+    uint64_t             extPanId;
+    std::vector<uint8_t> masterKey;
+    std::vector<uint8_t> pskc;
+    uint32_t             channelMask;
+
+    auto args = std::tie(masterKey, panid, name, extPanId, pskc, channelMask);
+
+    if (DBusMessageToTuple(*aRequest.GetMessage(), args) != OTBR_ERROR_NONE)
+    {
+        aRequest.ReplyOtbrResult(OTBR_OT_ERROR_INVALID_ARGS);
+    }
+    else
+    {
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+        extPanId = bswap_64(extPanId);
+#endif
+        threadHelper->Attach(name, panid, extPanId, masterKey, pskc, channelMask, [aRequest](otError aError) mutable {
+            aRequest.ReplyOtbrResult(static_cast<otbrError>(aError));
+        });
+    }
+}
+
+void DBusThreadObject::FactoryResetHandler(DBusRequest &aRequest)
+{
+    aRequest.ReplyOtbrResult(OTBR_ERROR_NONE);
+    mNcp->Reset();
+
+    {
+        auto threadHelper = mNcp->GetThreadHelper();
+        threadHelper->AddDeviceRoleHandler(std::bind(&DBusThreadObject::DeviceRoleHandler, this, _1));
+        SignalPropertyChanged(OTBR_DBUS_PREFIX, OTBR_DBUS_DEVICE_ROLE_PROPERTY,
+                              GetDeviceRoleName(OT_DEVICE_ROLE_DISABLED));
+    }
+}
+
+void DBusThreadObject::ResetHandler(DBusRequest &aRequest)
+{
+    auto threadHelper = mNcp->GetThreadHelper();
+
+    otInstanceReset(threadHelper->GetInstance());
+    aRequest.ReplyOtbrResult(OTBR_ERROR_NONE);
+}
+
+void DBusThreadObject::JoinerStartHandler(DBusRequest &aRequest)
+{
+    auto        threadHelper = mNcp->GetThreadHelper();
+    std::string pskd, provisionUrl, vendorName, vendorModel, vendorSwVersion, vendorData;
+    auto        args = std::tie(pskd, provisionUrl, vendorName, vendorModel, vendorSwVersion, vendorData);
+
+    if (DBusMessageToTuple(*aRequest.GetMessage(), args) != OTBR_ERROR_NONE)
+    {
+        aRequest.ReplyOtbrResult(OTBR_OT_ERROR_INVALID_ARGS);
+    }
+    else
+    {
+        threadHelper->JoinerStart(pskd, provisionUrl, vendorName, vendorModel, vendorSwVersion, vendorData,
+                                  [aRequest](otError aError) mutable { aRequest.ReplyOtbrResult(OTBR_ERROR_NONE); });
+    }
+}
+
+void DBusThreadObject::JoinerStopHandler(DBusRequest &aRequest)
+{
+    auto threadHelper = mNcp->GetThreadHelper();
+
+    otJoinerStop(threadHelper->GetInstance());
+    aRequest.ReplyOtbrResult(OTBR_ERROR_NONE);
+}
+
+void DBusThreadObject::OpenUnsecurePortHandler(DBusRequest &aRequest)
+{
+    auto     threadHelper = mNcp->GetThreadHelper();
+    uint16_t port;
+    uint32_t timeout;
+    auto     args = std::tie(port, timeout);
+
+    if (DBusMessageToTuple(*aRequest.GetMessage(), args) != OTBR_ERROR_NONE)
+    {
+        aRequest.ReplyOtbrResult(OTBR_OT_ERROR_INVALID_ARGS);
+    }
+    else
+    {
+        aRequest.ReplyOtbrResult(static_cast<otbrError>(threadHelper->OpenUnsecurePort(port, timeout)));
```

Comment:
Don't we need to handle endianness of `port` and `timeout`?

Reply:
See, it is error-prone.

Reply:
As explained before, only uint64_t has endianness problems.

Reply:
oh, it seems D-Bus APIs have already handled endianness for user.